karapace/utils.py

```diff
@@ -70,18 +80,53 @@ def default_json_serialization(
     assert_never(f"Object of type {obj.__class__.__name__!r} is not JSON serializable")
 
 
-SEPARATORS = (",", ":")
+@overload
+def json_encode(
+    obj: Any,
+    *,
+    binary: bool = False,
+    sort_keys: Optional[bool] = None,
+    compact: Optional[bool] = None,
+    indent: Optional[int] = None,
+) -> str:
+    ...
+
+
+@overload
+def json_encode(
+    obj: Any,
+    *,
+    binary: bool = True,
+    sort_keys: Optional[bool] = None,
+    compact: Optional[bool] = None,
+    indent: Optional[int] = None,
+) -> bytes:
+    ...
 
 
-def json_encode(obj, *, sort_keys: bool = True, binary=False, compact=False):
-    separators = SEPARATORS if compact else None
-    res = json.dumps(
-        obj,
-        sort_keys=sort_keys,
-        default=default_json_serialization,
-        separators=separators,
-    )
-    return res.encode("utf-8") if binary else res
+def json_encode(
+    obj: Any,
+    *,
+    binary: bool = False,
+    sort_keys: Optional[bool] = None,
+    compact: Optional[bool] = None,
+    indent: Optional[int] = None,
+) -> AnyStr:
+    kwargs = {}
+    if indent is not None:
+        kwargs["indent"] = indent
+    if compact is not False and indent is None:
+        kwargs["separators"] = (",", ":")
+    if sort_keys is True:
+        kwargs["sort_keys"] = True
+    result = json.dumps(obj, default=default_json_serialization, **kwargs)
+    return result.encode("utf8") if binary is True else result
+
+
+def json_decode(content: [AnyStr, IO[AnyStr]]) -> JsonData:
+    if isinstance(content, str) or isinstance(content, bytes):  # pylint: disable=consider-merging-isinstance
```

Comment:
Why would we not follow pylint's suggestion? It is more efficient to perform the work in C.

```suggestion
    if isinstance(content, (str, bytes)):
```

Reply:
Not possible in Python 3.7